tf_quant_finance/black_scholes/vanilla_prices.py

```diff
@@ -182,6 +182,346 @@ def option_price(*,
+      """
+      out_map = down_and_in_call_tensor_greater_strike # Default
+      if tf.math.less(params[1], params[0]):
+        if tf.equal(params[2], down_and_in_call):
+          out_map = down_and_in_call_tensor_lower_strike
+        if tf.equal(params[2], down_and_in_put):
+          out_map = down_and_in_put_tensor_lower_strike
+        if tf.equal(params[2], up_and_in_call):
+          out_map = up_and_in_call_tensor_lower_strike
+        if tf.equal(params[2], up_and_in_put):
+          out_map = up_and_in_put_tensor_lower_strike
+        if tf.equal(params[2], down_and_out_call):
+          out_map = down_and_out_call_tensor_lower_strike
+        if tf.equal(params[2], down_and_out_put):
+          out_map = down_and_out_put_tensor_lower_strike
+        if tf.equal(params[2], up_and_out_call):
+          out_map = up_and_out_call_tensor_lower_strike
+        if tf.equal(params[2], up_and_out_put):
+          out_map = up_and_out_put_tensor_lower_strike
+      else:
+        if tf.equal(params[2], down_and_in_call):
+          out_map = down_and_in_call_tensor_greater_strike
+        if tf.equal(params[2], down_and_in_put):
+          out_map = down_and_in_put_tensor_greater_strike
+        if tf.equal(params[2], up_and_in_call):
+          out_map = up_and_in_call_tensor_greater_strike
+        if tf.equal(params[2], up_and_in_put):
+          out_map = up_and_in_put_tensor_greater_strike
+        if tf.equal(params[2], down_and_out_call):
+          out_map = down_and_out_call_tensor_greater_strike
+        if tf.equal(params[2], down_and_out_put):
+          out_map = down_and_out_put_tensor_greater_strike
+        if tf.equal(params[2], up_and_out_call):
+          out_map = up_and_out_call_tensor_greater_strike
+        if tf.equal(params[2], up_and_out_put):
+          out_map = up_and_out_put_tensor_greater_strike
+      return out_map
+    
+    @tf.function
+    def get_below_or_above(x):
+      if tf.math.less(x[0], x[1]):
+        return -1.
+      return 1.
+
+    @tf.function
+    def get_call_or_put(x):
+      if tf.equal(tf.math.mod(x, 2), 0):
+        return -1.
+      return 1.
+
+    # build call or put, below or above barrier price, and masks
+    if otype.shape == ():
+      below_or_above = get_below_or_above((asset_price, barrier_price))
+      call_or_put = get_call_or_put(otype)
+      masks = get_out_map((barrier_price, strike_price, otype))
+    else:
+      masks = tf.map_fn(
+          get_out_map,
+          (barrier_price, strike_price, otype), dtype=tf.float32)
+      below_or_above = tf.map_fn(
+          get_below_or_above,
+          (asset_price, barrier_price), dtype=tf.float32)
+      call_or_put = tf.map_fn(
+          get_call_or_put,
+          otype, dtype=tf.float32)
+
+    # Calculate params for integrals
+    time_asset_volatility = tf.math.multiply(
+        asset_volatility, tf.math.sqrt(time_to_maturity),
+        name="time_asset_volatility")
+    mu = tf.math.subtract(
```

Comment:
Please feel free to use the overloaded infix operators (it is generally preferred to use those where available). For example, a - b vs tf.math.subtract(a, b)

Reply:
Done!